### PR TITLE
Feature/gu client rpc

### PR DESCRIPTION
**Pull request by prekucki**

### CODE REVIEW

File: gu-actix/src/pipe.rs [hunk 1]
```diff
@@ -36,9 +37,10 @@ impl<T, E> Stream for AsyncReader<T, E> {
 impl io::Read for SyncReader<Bytes, io::Error> {
     fn read(&mut self, buf: &mut [u8]) -> Result<usize, io::Error> {
         if self.buffer.is_none() {
-            let is_full = self.rx.is_full();
+            let is_full = self.rx.as_ref().unwrap().is_full();
```
Comment (tworec) [inline gu-actix/src/pipe.rs]: I had removed this y-day after your review for cleanup... it seems not to have any side effects. so pls remove it also here

File: gu-actix/src/release.rs [hunk 2]
```diff
@@ -208,7 +208,7 @@ mod test {
         });
 
         use std::thread::sleep;
-        sleep(Duration::from_secs(5));
+        sleep(Duration::from_secs(1));
```
Comment (tworec) [inline gu-actix/src/release.rs:212]: why do we need 1 s here?
furthermore, few lines above you are using sleep from tokio_timer. Can't it be used here also?

File: gu-provider/src/dockerman.rs [hunk 3]
```diff
@@ -548,9 +548,9 @@ impl Handler<SessionUpdate> for DockerMan {
 
     fn handle(&mut self, msg: SessionUpdate, _ctx: &mut Self::Context) -> Self::Result {
         if !self.deploys.contains_deploy(&msg.session_id) {
-            return ActorResponse::reply(Err(
-                vec![Error::NoSuchSession(msg.session_id).to_string()],
-            ));
+            return ActorResponse::reply(Err(vec![
```
Comment (tworec) [inline gu-provider/src/dockerman.rs]:
```suggestion
            return ActorResponse::reply(Err([
```

File: gu-provider/src/hdman.rs [hunk 4]
```diff
@@ -460,9 +460,9 @@ impl Handler<SessionUpdate> for HdMan {
 
     fn handle(&mut self, msg: SessionUpdate, _ctx: &mut Self::Context) -> Self::Result {
         if !self.deploys.contains_deploy(&msg.session_id) {
-            return ActorResponse::reply(Err(
-                vec![Error::NoSuchSession(msg.session_id).to_string()],
-            ));
+            return ActorResponse::reply(Err(vec![
```
Comment (tworec) [inline gu-provider/src/hdman.rs]:
```suggestion
            return ActorResponse::reply(Err([
```

File: gu-client/src/async.rs [hunk 5]
```diff
@@ -661,4 +691,34 @@ impl ProviderRef {
...
+        T::Result: DeserializeOwned,
+    {
+        let url = format!(
+            "{}peers/send-to/{:?}/{}",
+            self.connection.url(),
+            self.node_id,
+            T::ID
+        );
+
+        let hub_connection = self.clone();
```
Comment (tworec) [inline gu-client/src/async.rs]: unused stmnt

**Review by tworec — APPROVED** [review]
looks good, just minor comments

File: gu-client/src/lib.rs [hunk 6]
```diff
@@ -7,6 +7,8 @@ pub mod sync;
 pub use gu_model as model;
 pub use gu_net::NodeId;
 
+pub use gu_net::rpc::PublicMessage;
+
 #[cfg(feature = "integration_tests")]
 #[cfg(test)]
 mod integration_tests;
```
Comment (tworec) [inline gu-client/src/lib.rs:14]: those integration tests fails on mac

**Comment by marmistrz** [comment]
4470330d doesn't compile for me (after running `cargo update`)
```
error: cannot find derive macro `Serialize` in this scope
   --> /home/marcin/.cargo/git/checkouts/golem-unlimited-8aa5a0012b75d98f/4470330/gu-client/src/async.rs:720:10
    |
720 | #[derive(Serialize)]
    |          ^^^^^^^^^

error[E0277]: the trait bound `async::Body<T>: serde::ser::Serialize` is not satisfied
   --> /home/marcin/.cargo/git/checkouts/golem-unlimited-8aa5a0012b75d98f/4470330/gu-client/src/async.rs:711:14
    |
711 |             .json(Body { b: msg })
    |              ^^^^ the trait `serde::ser::Serialize` is not implemented for `async::Body<T>`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
error: Could not compile `gu-client`.
```

File: gu-actix/src/pipe.rs (hunk 1 above)
Comment (prekucki) [reply gu-actix/src/pipe.rs]: i put comment there. about this.

File: gu-actix/src/release.rs (hunk 2 above)
Comment (prekucki) [reply gu-actix/src/release.rs:212]: Q: "why do we need 1 s here?" 
A: it is a test ? sleep size is arbitrary. 
Q: " above you are using sleep from tokio_timer. Can't it be used here also?"
A: pls. read this test again. what we testing ? whats is diffrerence betweem this places.

File: gu-client/src/lib.rs (hunk 6 above)
Comment (prekucki) [reply gu-client/src/lib.rs:14]: :rofl: they always fail. you should run then from tests/Makefile only (inside docker container)!

Comment (tworec) [reply gu-client/src/lib.rs:14]: I'm running with `--all-features` so thats why it was invoked.

**Comment by tworec** [comment]
@marmistrz this is fixed in #213

**Pull request merged**